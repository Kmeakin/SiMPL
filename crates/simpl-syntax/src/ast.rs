pub type Ident = String;

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit {
        val: Lit,
    },
    Var {
        name: String,
    },
    If {
        test: Box<Self>,
        then_branch: Box<Self>,
        else_branch: Box<Self>,
    },
    Let {
        bindings: Vec<(Ident, Self)>,
        body: Box<Self>,
    },
    Letrec {
        bindings: Vec<(Ident, Self)>,
        body: Box<Self>,
    },
    Lambda {
        params: Vec<Ident>,
        body: Box<Self>,
    },
    App {
        func: Box<Self>,
        arg: Box<Self>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum Lit {
    Bool(bool),
    Int(i64),
    Float(f64),
}
