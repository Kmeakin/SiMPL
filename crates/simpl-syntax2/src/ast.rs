pub type Ident = String;

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit {
        val: Lit,
    },
    Var {
        name: String,
    },
    If {
        test: Box<Expr>,
        then_branch: Box<Expr>,
        else_branch: Box<Expr>,
    },
    Let {
        bindings: Vec<(Ident, Expr)>,
        body: Box<Expr>,
    },
    Lambda {
        params: Vec<Ident>,
        body: Box<Expr>,
    },
    App {
        func: Box<Expr>,
        arg: Box<Expr>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum Lit {
    Bool(bool),
    Int(i64),
    Float(f64),
}
