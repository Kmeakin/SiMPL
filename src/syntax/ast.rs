pub use crate::types::ty::Type;
use derive_more::Display;
pub use simple_symbol::Symbol;

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit {
        val: Lit,
    },
    Var {
        name: Symbol,
    },
    Binop {
        lhs: Box<Self>,
        rhs: Box<Self>,
        op: Binop,
    },
    If {
        test: Box<Self>,
        then: Box<Self>,
        els: Box<Self>,
    },
    Let {
        bindings: Vec<LetBinding>,
        body: Box<Self>,
    },
    Letrec {
        bindings: Vec<LetBinding>,
        body: Box<Self>,
    },
    Lambda {
        params: Vec<Param>,
        body: Box<Self>,
    },
    App {
        func: Box<Self>,
        arg: Box<Self>,
    },
}

#[derive(Debug, Clone, PartialEq, Display)]
pub enum Lit {
    #[display(fmt = "{}", _0)]
    Bool(bool),
    #[display(fmt = "{}", _0)]
    Int(i64),
    #[display(fmt = "{}", _0)]
    Float(f64),
}

#[derive(Debug, Clone, PartialEq)]
pub struct LetBinding {
    pub name: Symbol,
    pub ann: Option<Type>,
    pub val: Box<Expr>,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum Binop {
    IntAdd,
    IntSub,
    IntMul,
    IntDiv,
    IntLt,
    IntLeq,
    IntGt,
    IntGeq,
    FloatAdd,
    FloatSub,
    FloatMul,
    FloatDiv,
    FloatLt,
    FloatLeq,
    FloatGt,
    FloatGeq,
    Eq,
    Neq,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    pub name: Symbol,
    pub ann: Option<Type>,
}

impl From<bool> for Lit {
    fn from(other: bool) -> Self {
        Self::Bool(other)
    }
}

impl From<i64> for Lit {
    fn from(other: i64) -> Self {
        Self::Int(other)
    }
}

impl From<f64> for Lit {
    fn from(other: f64) -> Self {
        Self::Float(other)
    }
}
