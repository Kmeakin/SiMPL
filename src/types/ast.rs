pub use crate::syntax::ast::{Ident, Lit};
use crate::{
    syntax::ast::Expr,
    types::ty::{Type, TypeEnv},
    util::counter::{Counter, FromId},
};
use std::str::FromStr;

#[derive(Debug, Clone, PartialEq)]
pub enum TypedExpr {
    Lit {
        ty: Type,
        val: Lit,
    },
    Var {
        ty: Type,
        name: String,
    },
    If {
        ty: Type,
        test: Box<Self>,
        then_branch: Box<Self>,
        else_branch: Box<Self>,
    },
    Let {
        ty: Type,
        binding: LetBinding,
        body: Box<Self>,
    },
    Letrec {
        ty: Type,
        bindings: Vec<LetBinding>,
        body: Box<Self>,
    },
    Lambda {
        ty: Type,
        param: Param,
        body: Box<Self>,
    },
    App {
        ty: Type,
        func: Box<Self>,
        arg: Box<Self>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    pub ty: Type,
    pub name: Ident,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LetBinding {
    pub ty: Type,
    pub name: Ident,
    pub val: Box<TypedExpr>,
}

pub type TypeVarGen = Counter<Type>;
impl FromId for Type {
    fn from_id(id: u32) -> Self {
        Self::Var(id)
    }
}

impl FromStr for TypedExpr {
    type Err = String;
    fn from_str(src: &str) -> Result<Self, Self::Err> {
        // TODO: return a trait object instead of unwrapping
        let ast = crate::syntax::parse(src).unwrap();
        Self::from_ast(ast)
    }
}

impl TypedExpr {
    pub fn from_ast(ast: Expr) -> Result<Self, String> {
        let mut gen = TypeVarGen::new();
        let tenv = TypeEnv::default();
        Self::from_ast_inner(ast, &tenv, &mut gen)
    }

    fn from_ast_inner(ast: Expr, tenv: &TypeEnv, gen: &mut TypeVarGen) -> Result<Self, String> {
        let expr = match ast {
            Expr::Lit { val } => Self::Lit {
                val,
                ty: gen.next(),
            },
            Expr::Var { name } => match tenv.get(&name) {
                None => return Err(format!("Unbound variable: {}", name)),
                Some(ty) => Self::Var {
                    ty: ty.clone(),
                    name,
                },
            },
            Expr::If {
                test,
                then_branch,
                else_branch,
            } => Self::If {
                ty: gen.next(),
                test: box Self::from_ast_inner(*test, tenv, gen)?,
                then_branch: box Self::from_ast_inner(*then_branch, tenv, gen)?,
                else_branch: box Self::from_ast_inner(*else_branch, tenv, gen)?,
            },
            Expr::Let { bindings, body } => {
                let (bindings, body) = expand_let(&bindings, *body);

                assert!(bindings.len() == 1);
                let ty = gen.next();

                let (binding_name, binding_val) = &bindings[0];
                let binding_ty = gen.next();
                let mut extended_tenv = tenv.clone();
                extended_tenv.insert(binding_name.into(), binding_ty.clone());

                Self::Let {
                    ty,
                    binding: LetBinding {
                        ty: binding_ty,
                        name: binding_name.clone(),
                        val: box Self::from_ast_inner(binding_val.clone(), tenv, gen)?,
                    },
                    body: box Self::from_ast_inner(body, &extended_tenv, gen)?,
                }
            }
            Expr::Letrec { bindings, body } => {
                let ty = gen.next();
                let mut extended_tenv = tenv.clone();

                let mut tvars = vec![];
                for (name, _) in &bindings {
                    let tvar = gen.next();
                    extended_tenv.insert(name.clone(), tvar.clone());
                    tvars.push(tvar);
                }

                let mut new_bindings = vec![];

                for ((name, val), tvar) in bindings.into_iter().zip(tvars) {
                    new_bindings.push(LetBinding {
                        name: name.clone(),
                        ty: tvar,
                        val: box Self::from_ast_inner(val.clone(), &extended_tenv, gen)?,
                    });
                }

                Self::Letrec {
                    ty,
                    bindings: new_bindings,
                    body: box Self::from_ast_inner(*body, &extended_tenv, gen)?,
                }
            }
            Expr::Lambda { params, body } => {
                let (params, body) = expand_lambda(&params, *body);

                assert!(params.len() == 1);
                let ty = gen.next();

                let param_name = &params[0];
                let param_ty = gen.next();
                let mut extended_tenv = tenv.clone();
                extended_tenv.insert(param_name.clone(), param_ty.clone());
                Self::Lambda {
                    ty,
                    param: Param {
                        ty: param_ty,
                        name: param_name.clone(),
                    },
                    body: box Self::from_ast_inner(body, &extended_tenv, gen)?,
                }
            }
            Expr::App { func, arg } => Self::App {
                ty: gen.next(),
                func: box Self::from_ast_inner(*func, tenv, gen)?,
                arg: box Self::from_ast_inner(*arg, tenv, gen)?,
            },
        };
        Ok(expr)
    }

    pub fn ty(&self) -> Type {
        match self {
            Self::Lit { ty, .. }
            | Self::Var { ty, .. }
            | Self::If { ty, .. }
            | Self::Let { ty, .. }
            | Self::Letrec { ty, .. }
            | Self::Lambda { ty, .. }
            | Self::App { ty, .. } => ty.clone(),
        }
    }
}

/// Expand an `ast::Expr::Let` with many bound variables into a nested
/// sequences of lets each binding a single variable
/// eg `let x = 1, y = 2 in add x y` expands to
///    `let x = 1 in let y = 2 in add x y`
fn expand_let(bindings: &[(Ident, Expr)], body: Expr) -> (Vec<(Ident, Expr)>, Expr) {
    assert!(!bindings.is_empty());
    if bindings.len() == 1 {
        (bindings.to_vec(), body)
    } else {
        let binding = &bindings[0];

        let (rest_bindings, rest_body) = expand_let(&bindings[1..], body);

        (
            vec![binding.clone()],
            Expr::Let {
                bindings: rest_bindings,
                body: box rest_body,
            },
        )
    }
}

/// Expand an `ast::Expr::Lambda` with many parameters into a nested
/// sequences of lambdas each binding a single parameter
/// eg `\x, y -> add x y` expands to
///    `\x -> \y -> add x y`
fn expand_lambda(params: &[Ident], body: Expr) -> (Vec<Ident>, Expr) {
    assert!(!params.is_empty());
    if params.len() == 1 {
        (params.to_vec(), body)
    } else {
        let param = &params[0];

        let (rest_params, rest_body) = expand_lambda(&params[1..], body);

        (
            vec![param.clone()],
            Expr::Lambda {
                params: rest_params,
                body: box rest_body,
            },
        )
    }
}

#[test]
fn test_expand_let() {
    let bindings = vec![
        ("x".into(), Expr::Lit { val: Lit::Int(1) }),
        ("y".into(), Expr::Lit { val: Lit::Int(2) }),
    ];

    let body = Expr::Lit { val: Lit::Int(0) };

    assert_eq!(
        expand_let(&bindings, body),
        (
            vec![("x".into(), Expr::Lit { val: Lit::Int(1) }),],
            Expr::Let {
                bindings: vec![("y".into(), Expr::Lit { val: Lit::Int(2) })],
                body: box Expr::Lit { val: Lit::Int(0) }
            }
        ),
    )
}

#[test]
fn test_expand_lambda() {
    let params = vec!["x".into(), "y".into()];
    let body = Expr::Lit { val: Lit::Int(0) };

    assert_eq!(
        expand_lambda(&params, body),
        (
            vec!["x".into()],
            Expr::Lambda {
                params: vec!["y".into()],
                body: box Expr::Lit { val: Lit::Int(0) }
            }
        ),
    )
}
