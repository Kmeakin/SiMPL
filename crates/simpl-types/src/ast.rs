use crate::{
    ty::{Type, TypeEnv},
    FromId, IdGen,
};
use simpl_syntax::ast::Expr;
pub use simpl_syntax::ast::{Ident, Lit};

#[derive(Debug, Clone, PartialEq)]
pub enum TypedExpr {
    Lit {
        ty: Type,
        val: Lit,
    },
    Var {
        ty: Type,
        name: String,
    },
    If {
        ty: Type,
        test: Box<Self>,
        then_branch: Box<Self>,
        else_branch: Box<Self>,
    },
    Let {
        ty: Type,
        binding: LetBinding,
        body: Box<Self>,
    },
    Letrec {
        ty: Type,
        bindings: Vec<LetBinding>,
        body: Box<Self>,
    },
    Lambda {
        ty: Type,
        param: Param,
        body: Box<Self>,
    },
    App {
        ty: Type,
        func: Box<Self>,
        arg: Box<Self>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    pub ty: Type,
    pub name: Ident,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LetBinding {
    pub ty: Type,
    pub name: Ident,
    pub val: Box<TypedExpr>,
}

pub type TypeVarGen = IdGen<Type>;
impl FromId for Type {
    fn from_id(id: u32) -> Type {
        Type::Var(id)
    }
}

impl TypedExpr {
    pub fn from_str(src: &str) -> Result<Self, String> {
        // TODO: return a trait object instead of unwrapping
        let ast = simpl_syntax::parse(src).unwrap();
        Self::from_ast(ast)
    }

    pub fn from_ast(ast: Expr) -> Result<Self, String> {
        let mut gen = TypeVarGen::new();
        let tenv = TypeEnv::default();
        Self::from_ast_inner(ast, &tenv, &mut gen)
    }

    fn from_ast_inner(ast: Expr, tenv: &TypeEnv, gen: &mut TypeVarGen) -> Result<Self, String> {
        let expr = match ast {
            Expr::Lit { val } => Self::Lit {
                val,
                ty: gen.next(),
            },
            Expr::Var { name } => match tenv.get(&name) {
                None => return Err(format!("Unbound variable: {}", name)),
                Some(ty) => Self::Var {
                    ty: ty.clone(),
                    name,
                },
            },
            Expr::If {
                test,
                then_branch,
                else_branch,
            } => Self::If {
                ty: gen.next(),
                test: box Self::from_ast_inner(*test, tenv, gen)?,
                then_branch: box Self::from_ast_inner(*then_branch, tenv, gen)?,
                else_branch: box Self::from_ast_inner(*else_branch, tenv, gen)?,
            },
            Expr::Let { bindings, body } => {
                assert!(bindings.len() >= 1);
                let ty = gen.next();

                // TODO: desugar `let v1 = e1, v2 = e2 in b` into
                // `let v1 = e1 in let v2 = e2 in b`
                let (binding_name, binding_val) = &bindings[0];
                let binding_ty = gen.next();
                let mut extended_tenv = tenv.clone();
                extended_tenv.insert(binding_name.into(), binding_ty.clone());

                Self::Let {
                    ty,
                    binding: LetBinding {
                        ty: binding_ty,
                        name: binding_name.clone(),
                        val: box Self::from_ast_inner(binding_val.clone(), tenv, gen)?,
                    },
                    body: box Self::from_ast_inner(*body, &extended_tenv, gen)?,
                }
            }
            Expr::Letrec { bindings, body } => {
                let ty = gen.next();
                let mut extended_tenv = tenv.clone();

                let mut tvars = vec![];
                for (name, _) in &bindings {
                    let tvar = gen.next();
                    extended_tenv.insert(name.clone(), tvar.clone());
                    tvars.push(tvar);
                }

                let mut new_bindings = vec![];

                for ((name, val), tvar) in bindings.into_iter().zip(tvars) {
                    new_bindings.push(LetBinding {
                        name: name.clone(),
                        ty: tvar,
                        val: box Self::from_ast_inner(val.clone(), &extended_tenv, gen)?,
                    });
                }

                Self::Letrec {
                    ty,
                    bindings: new_bindings,
                    body: box Self::from_ast_inner(*body, &extended_tenv, gen)?,
                }
            }
            Expr::Lambda { params, body } => {
                assert!(params.len() >= 1);
                let ty = gen.next();

                // TODO: desugar `\x, y -> e` into `\x -> \y -> e`
                let param_name = &params[0];
                let param_ty = gen.next();
                let mut extended_tenv = tenv.clone();
                extended_tenv.insert(param_name.clone(), param_ty.clone());
                Self::Lambda {
                    ty,
                    param: Param {
                        ty: param_ty,
                        name: param_name.clone(),
                    },
                    body: box Self::from_ast_inner(*body, &extended_tenv, gen)?,
                }
            }
            Expr::App { func, arg } => Self::App {
                ty: gen.next(),
                func: box Self::from_ast_inner(*func, tenv, gen)?,
                arg: box Self::from_ast_inner(*arg, tenv, gen)?,
            },
        };
        Ok(expr)
    }

    pub fn ty(&self) -> Type {
        match self {
            Self::Lit { ty, .. }
            | Self::Var { ty, .. }
            | Self::If { ty, .. }
            | Self::Let { ty, .. }
            | Self::Letrec { ty, .. }
            | Self::Lambda { ty, .. }
            | Self::App { ty, .. } => ty.clone(),
        }
    }
}
