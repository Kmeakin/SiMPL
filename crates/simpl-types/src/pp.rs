use crate::ast::TypedExpr as Expr;
use pretty::{Doc, RcDoc};

const INDENT: isize = 4;
const WIDTH: usize = 40;

impl Expr {
    pub fn to_doc(&self) -> RcDoc<()> {
        match self {
            Self::Lit { ty, val } => RcDoc::as_string(val),
            Self::Var { ty, name } => RcDoc::as_string(name),
            Self::If {
                ty,
                test,
                then_branch,
                else_branch,
            } => RcDoc::text("if")
                .append(RcDoc::space())
                .append(test.to_doc())
                .append(RcDoc::line())
                .append(RcDoc::text("then"))
                .append(RcDoc::space())
                .append(then_branch.to_doc())
                .append(RcDoc::line())
                .append(RcDoc::text("else"))
                .append(RcDoc::space())
                .append(else_branch.to_doc())
                .nest(INDENT)
                .group(),
            Self::App { ty, func, arg } => RcDoc::text("(")
                .append(func.to_doc())
                .append(RcDoc::space())
                .append(arg.to_doc())
                .append(RcDoc::text(")"))
                .group(),
            Self::Lambda { ty, param, body } => RcDoc::text(r"\")
                .append(param.name.clone())
                .append(RcDoc::space())
                .append(RcDoc::text("->"))
                .append(RcDoc::space())
                .append(body.to_doc())
                .group(),
            Self::Let { ty, binding, body } => RcDoc::text("let")
                .append(RcDoc::space())
                .append(binding.name.clone())
                .append(RcDoc::space())
                .append(RcDoc::text("="))
                .append(RcDoc::space())
                .append(binding.val.to_doc())
                .append(RcDoc::space())
                .append(RcDoc::text("in"))
                .append(RcDoc::line())
                .append(body.to_doc())
                .nest(INDENT)
                .group(),
            Self::Letrec { ty, bindings, body } => todo!(),
        }
    }

    pub fn pretty(&self) -> String {
        let mut v = Vec::new();
        self.to_doc().render(WIDTH, &mut v).unwrap();
        String::from_utf8(v).unwrap()
    }
}

mod test {
    use super::*;
    use crate::parse_and_type;
    use insta::assert_snapshot;

    #[track_caller]
    fn test_pp(src: &str) {
        let expr = crate::parse_and_type(src);
        assert_snapshot!(expr.pretty());
    }

    #[test]
    fn pp_lit() {
        test_pp("123");
    }

    #[test]
    fn pp_var() {
        test_pp("add");
    }

    #[test]
    fn pp_if() {
        test_pp("if true then 1 else 0");
        test_pp("if (if false then true else true) then (if false then 1 else 0) else (if false then 420 else 69)");
    }

    #[test]
    fn pp_app() {
        test_pp("if not false then 1 else 0");
        test_pp("if not (is_zero (add 1 1)) then 50 else 100");
    }

    #[test]
    fn pp_lambda() {
        test_pp(r"if not false then \x -> x else \x -> not x");
        test_pp(r"if not false then \a, b -> a else \x, y -> y");
    }

    #[test]
    fn pp_let() {
        test_pp(r"let x = 5 in x");
        test_pp(r"let id = \x -> x, first = \a, b -> a in id not (first true 1)");
    }
}
