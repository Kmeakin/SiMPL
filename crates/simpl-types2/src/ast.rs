use crate::{
    ty::{Type, TypeEnv},
    FromId, IdGen,
};
use simpl_syntax2::ast::Expr;
pub use simpl_syntax2::ast::{Ident, Lit};

#[derive(Debug, Clone, PartialEq)]
pub enum TypedExpr {
    Lit {
        ty: Type,
        val: Lit,
    },
    Var {
        ty: Type,
        name: String,
    },
    If {
        ty: Type,
        test: Box<Self>,
        then_branch: Box<Self>,
        else_branch: Box<Self>,
    },
    Let {
        ty: Type,
        binding: (Type, Ident, Box<Self>),
        body: Box<Self>,
    },
    Letrec {
        ty: Type,
        bindings: Vec<(Type, Ident, Self)>,
        body: Box<Self>,
    },
    Lambda {
        ty: Type,
        param: (Type, Ident),
        body: Box<Self>,
    },
    App {
        ty: Type,
        func: Box<Self>,
        arg: Box<Self>,
    },
}

pub type TypeVarGen = IdGen<Type>;
impl FromId for Type {
    fn from_id(id: u32) -> Type {
        Type::Var(id)
    }
}

impl TypedExpr {
    pub fn from_ast(ast: Expr) -> Result<Self, String> {
        let mut gen = TypeVarGen::new();
        let tenv = TypeEnv::default();
        Self::from_ast_inner(ast, &tenv, &mut gen)
    }

    fn from_ast_inner(ast: Expr, tenv: &TypeEnv, gen: &mut TypeVarGen) -> Result<Self, String> {
        let expr = match ast {
            Expr::Lit { val } => Self::Lit {
                val,
                ty: gen.next(),
            },
            Expr::Var { name } => match tenv.get(&name) {
                None => return Err(format!("Unbound variable: {}", name)),
                Some(ty) => Self::Var {
                    ty: ty.clone(),
                    name,
                },
            },
            Expr::If {
                test,
                then_branch,
                else_branch,
            } => Self::If {
                ty: gen.next(),
                test: box Self::from_ast_inner(*test, tenv, gen)?,
                then_branch: box Self::from_ast_inner(*then_branch, tenv, gen)?,
                else_branch: box Self::from_ast_inner(*else_branch, tenv, gen)?,
            },
            Expr::Let { bindings, body } => todo!(),
            Expr::Letrec { bindings, body } => todo!(),
            Expr::Lambda { params, body } => todo!(),
            Expr::App { func, arg } => Self::App {
                ty: gen.next(),
                func: box Self::from_ast_inner(*func, tenv, gen)?,
                arg: box Self::from_ast_inner(*arg, tenv, gen)?,
            },
        };
        Ok(expr)
    }

    pub fn ty(&self) -> Type {
        match self {
            Self::Lit { ty, .. }
            | Self::Var { ty, .. }
            | Self::If { ty, .. }
            | Self::Let { ty, .. }
            | Self::Letrec { ty, .. }
            | Self::Lambda { ty, .. }
            | Self::App { ty, .. } => ty.clone(),
        }
    }
}
