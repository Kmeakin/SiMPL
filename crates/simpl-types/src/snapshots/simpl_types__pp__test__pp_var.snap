---
source: crates/simpl-types/src/pp.rs
expression: expr.pretty()
---
add
